### ISSUE TITLE
Reimplement Qt specific codes to use Ginga's general widgets

### ISSUE THREAD

**Issue by pllim** [spacetelescope/stginga#62]
~~Whenever it is available under Anaconda, we have to make sure it works under both Qt4 and Qt5.~~

Move away from Qt-specific codes by using Ginga's toolkit-agnostic widgets. Then, we can stop requiring `stginga` to only work with Qt toolkit.

**Comment by ejeschke** [comment]
You might want to support pyside as well. IIRC I think @perrygreenfield liked their license better.  I'm not sure what state that project is in.

"Stock" ginga right now supports qt4/qt5/pyside (and gtk and (soon) web) so as long as you stick to the wrapped widgets you should be good.  That is another benefit of using the wrapped widgets: when a bug fix (or e.g. pyqt6) comes out you won't have to exhaustively check all your plugins to make sure they work under every possible scenario.

**Comment by pllim** [comment]
Thanks for the info, @ejeschke . 

Your wrapped widgets are nice but I don't know how to use them to implement the following:

``` python
b.background_value.widget.setReadOnly(True)
b.background_value.widget.setEnabled(True)
b.background_value.widget.setStyleSheet('QLineEdit{background: white;}')
```

What I am trying to do above is to set an editable text box such that user cannot edit it, but still can highlight the contents for copy and paste. Also I am overriding the default grey background so the value is easy to read.

**Comment by stscieisenhamer** [comment]
The pyside license is better. However, for quite a time, support had been lacking. However, recently, it has picked up again, now with Qt5 support. So, its looking like PySide is a player again.

**Comment by pllim** [comment]
But perhaps I'll change the issue title to something else, as suggested by @ejeschke .

**Comment by ejeschke** [comment]
@pllim, what kind of widget is `b.background_value`?  Is it a TextArea?  If so, you can use the `editable=False` in the constructor to prevent it from being edited.  It can then still be copied.

**Comment by ejeschke** [comment]
Wrapped widgets also use PEP8 api :-)

**Comment by pllim** [comment]
@ejeschke , it is a `QLineEdit` object, created from the following in the plugin:

``` python
captions = ((..., 'Background Value', 'entry'), )
w, b = Widgets.build_info(captions, orientation=self.orientation)
self.w.update(b)
```

**Comment by ejeschke** [comment]
Use `set_enabled(False)`.  Or `True`, as needed.

**Comment by pllim** [comment]
@ejeschke , I started with that, but it didn't give me the desired behavior. I probably will have to add a new method to your `TextEntry` class in `ginga/qtw/Widgets.py`, but then I'll break the other toolkits because I don't know how to implement, say, the GTK equivalent of what I did in Qt.

**Comment by ejeschke** [comment]
@pllim, try commit 5d0fe15e30b14fbf52356a713d038e9ffa28a70a from ginga master.  It adds the `set_editable` method and an `editable` parameter to the `TextEntry` widget.

Note: ejeschke/ginga@5d0fe15e3

**Comment by pllim** [comment]
@ejeschke , thanks for the quick fix! I'll let you know.

**Comment by pllim** [comment]
@ejeschke , how do I enable/disable a button using the agnostic widget? I don't see any implementation of `set_enable()` or such in `ginga/qtw/Widgets.py` for `Button` object.

**Comment by pllim** [comment]
@ejeschke , I also cannot find a way to do this in `TextEntry`:

``` python
b.background_value.widget.setStyleSheet('QLineEdit{background: white;}')
```

Basically, I just want to force the background not to grey out when the text is not editable.

**Comment by ejeschke** [comment]
`set_enabled(tf)` is defined in the base class, and will work for a Button.

**Comment by ejeschke** [comment]
What happens when you use `set_editable(False)` but `set_enabled(True)`?  Is it still grey?

**Comment by pllim** [comment]
Yeah, that works. Thanks!

**Comment by pllim** [comment]
I think only my plugins are affected by this. I have made the changes in #61.

**Comment by pllim** [comment]
So far, I found everything I need in the general Ginga widgets, except for ejeschke/ginga#256.

**Comment by pllim** [comment]
Fixed in #61
